### DISCUSSION TITLE
How to update python packages (pip install requirements.txt) periodically

### DISCUSSION THREAD

**Discussion by isaacnorman82** [apache/airflow#26674] [Q&A] [answered]
I can't see any good documentation on how to periodically update python packages in airflow for use by DAGs/Python operator.

If I have a kube helm based install I can have a Git sidecar.  I found this issue from a couple of years back that suggests a requirements.txt will be automatically used: https://github.com/helm/charts/issues/22744 
The comments there are out of date though, there's nothing about dags.initContainer anymore in the helm chart.  Also, even this seem like it'll be an update on first run rather than periodically (which I would think you would need if you're running git sync).

Taking a step back and considering the options laid out in https://airflow.apache.org/docs/helm-chart/stable/manage-dags-files.html:

1. Bake DAGs into your image - a single pip install is fine and so no issue here, but you have to update your image and redeploy every code change.
2. All other options are either DAGs from a mount/shared volume/Git side car and all amount to basically the same thing:  the code will update after the Airflow image is deployed.   Is there any method in this system for watching for a change and doing a pip install?

I could imagine running a python/bash operator on a manually or on a schedule that does a pip install, is this sensible?

**Comment by potiuk** [comment, marked as answer]
Nope. It is a very bad practice in Kubernetes/Container world. You should hav e pipeline where you build and push to your docker registry your own image (with updated requirements) and when it happens  deploy a helm chart switching to he new image when ready. This is the most reasonable and sustainable approach (And one that is "standard" in K8S world). You will find why dynamically installing package is a bad idea here: https://airflow.apache.org/docs/docker-stack/entrypoint.html#installing-additional-requirements
